airflow/dag_fetcher.py

```diff
@@ -0,0 +1,290 @@
+                        self.dagbag.import_errors[filepath] = str(e)
+                        self.dagbag.file_last_changed[filepath] = file_last_changed
+
+        for m in mods:
+            for dag in list(m.__dict__.values()):
+                if isinstance(dag, airflow.models.DAG):
+                    if not dag.full_filepath:
+                        dag.full_filepath = filepath
+                        if dag.fileloc != filepath:
+                            dag.fileloc = filepath
+                    try:
+                        dag.is_subdag = False
+                        self.dagbag.bag_dag(dag, parent_dag=dag, root_dag=dag)
+                        found_dags.append(dag)
+                        found_dags += dag.subdags
+                    except AirflowDagCycleException as cycle_exception:
+                        self.log.exception("Failed to bag_dag: %s", dag.full_filepath)
+                        self.dagbag.import_errors[dag.full_filepath] = \
+                            str(cycle_exception)
+                        self.dagbag.file_last_changed[dag.full_filepath] = \
+                            file_last_changed
+
+        self.dagbag.file_last_changed[filepath] = file_last_changed
+        return found_dags
+
+    def fetch(self, only_if_updated=True):
+        """
+        Walks the dags_folder (self.dags_uri) looking for files to process
+        """
+        if os.path.isfile(self.dags_uri):
+            self.process_file(self.dags_uri, only_if_updated=only_if_updated)
+        elif os.path.isdir(self.dags_uri):
+            patterns = []
+            for root, dirs, files in os.walk(self.dags_uri, followlinks=True):
+                ignore_file = [f for f in files if f == '.airflowignore']
+                if ignore_file:
+                    f = open(os.path.join(root, ignore_file[0]), 'r')
+                    patterns += [p for p in f.read().split('\n') if p]
+                    f.close()
+                for f in files:
+                    try:
+                        filepath = os.path.join(root, f)
+                        if not os.path.isfile(filepath):
+                            continue
+                        mod_name, file_ext = os.path.splitext(
+                            os.path.split(filepath)[-1])
+                        if file_ext != '.py' and not zipfile.is_zipfile(filepath):
+                            continue
+                        if not any(
+                                [re.findall(p, filepath) for p in patterns]):
+                            ts = timezone.utcnow()
+                            found_dags = self.process_file(
+                                filepath, only_if_updated=only_if_updated)
+
+                            td = timezone.utcnow() - ts
+                            td = td.total_seconds() + (
+                                float(td.microseconds) / 1000000)
+                            self.stats.append(self.FileLoadStat(
```

Comment:
self.stats should be called in process_file, otherwise if the uri is a single file, fetch won't be able to generate stats for it (I'm aware this is just copy/paste from old code, but just pointing it out)